packages/apps/app-dashboard/src/utils/user-dashboard/getAgentPKP.ts

```diff
@@ -1,52 +1,153 @@
 import { IRelayPKP } from '@lit-protocol/types';
 import { getPkpNftContract } from './get-pkp-nft-contract';
 import { SELECTED_LIT_NETWORK } from './lit';
+import { readOnlySigner } from '../developer-dashboard/readOnlySigner';
+import { getClient } from '@lit-protocol/vincent-contracts-sdk';
+
+export type AgentAppPermission = {
+  appId: number;
+  pkp: IRelayPKP;
+};
+
+export type GetAgentPKPsResult = {
+  permitted: AgentAppPermission[];
+  unpermitted: IRelayPKP[];
+};
 
 /**
- * Get Agent PKP for a user address
+ * Get Agent PKPs for a user address
  *
- * Finds an Agent PKP owned by the user that is different from their current PKP.
- * This is used for connect delegation and other agent-related operations.
+ * Finds all Agent PKPs owned by the user that are different from their current PKP.
+ * Separates them into permitted (flat array of app-PKP pairs) and unpermitted PKPs.
  *
  * @param userAddress The ETH address of the user's current PKP
- * @returns Promise<IRelayPKP> The Agent PKP details, or null if none found
- * @throws Error if no Agent PKP is found or if there's an issue with the contract calls
+ * @returns Promise<GetAgentPKPsResult> Object containing permitted and unpermitted PKPs
+ * @throws Error if there's an issue with the contract calls
  */
-export async function getAgentPKP(userAddress: string): Promise<IRelayPKP> {
+export async function getAgentPKPs(userAddress: string): Promise<GetAgentPKPsResult> {
   try {
     const pkpNftContract = getPkpNftContract(SELECTED_LIT_NETWORK);
 
     const balance = await pkpNftContract.balanceOf(userAddress);
     if (balance.toNumber() === 0) {
-      throw new Error('No PKPs found for this user');
+      return { permitted: [], unpermitted: [] };
     }
 
-    for (let i = 0; i < balance.toNumber(); i++) {
-      const tokenId = await pkpNftContract.tokenOfOwnerByIndex(userAddress, i);
-      const publicKey = await pkpNftContract.getPubkey(tokenId);
-      const ethAddress = await pkpNftContract.getEthAddress(tokenId);
+    // Get all token IDs in parallel first
+    const tokenIds = await Promise.all(
+      Array.from({ length: balance.toNumber() }, (_, i) =>
+        pkpNftContract.tokenOfOwnerByIndex(userAddress, i),
+      ),
+    );
 
-      if (ethAddress.toLowerCase() === userAddress.toLowerCase()) {
-        continue;
-      }
+    // Then get all PKP data in parallel
+    const pkpDataPromises = tokenIds.map(async (tokenId) => {
+      const [publicKey, ethAddress] = await Promise.all([
+        pkpNftContract.getPubkey(tokenId),
+        pkpNftContract.getEthAddress(tokenId),
```

Comment:
I think we could just derive the ETH address using [ethers.utils.computeAddress](https://docs.ethers.org/v5/api/utils/address/#utils-computeAddress) which would cut the number of RPC requests here in half